packages/liveblocks-react/src/room.tsx

```diff
@@ -1075,66 +1073,85 @@ export function createRoomContext<
   }
 
   function incrementQuerySubscribers(queryKey: string) {
-    const requestCache = requestsCache.get(queryKey);
+    const subscribers = subscribersByQuery.get(queryKey);
 
-    if (requestCache === undefined) {
+    if (subscribers === undefined) {
       console.warn(
         `Internal unexpected behavior. Cannot increase subscriber count for query "${queryKey}"`
       );
       return;
     }
 
-    requestCache.subscribers++;
+    subscribersByQuery.set(queryKey, subscribers + 1);
 
     poller.start(POLLING_INTERVAL);
   }
 
   function decrementQuerySubscribers(queryKey: string) {
-    const requestCache = requestsCache.get(queryKey);
+    const subscribers = subscribersByQuery.get(queryKey);
 
-    if (requestCache === undefined || requestCache.subscribers <= 0) {
+    if (subscribers === undefined || subscribers <= 0) {
       console.warn(
         `Internal unexpected behavior. Cannot decrease subscriber count for query "${queryKey}"`
       );
       return;
     }
 
-    requestCache.subscribers--;
+    subscribersByQuery.set(queryKey, subscribers - 1);
 
     let totalSubscribers = 0;
-    for (const requestCache of requestsCache.values()) {
-      totalSubscribers += requestCache.subscribers;
+    for (const subscribers of subscribersByQuery.values()) {
+      totalSubscribers += subscribers;
     }
 
     if (totalSubscribers <= 0) {
       poller.stop();
     }
   }
 
+  /**
+   * Retries an action using the exponential backoff algorithm
+   * @param action The action to retry
+   * @param retryCount The number of times the action has been retried
+   */
+  function handleError(action: () => void, retryCount: number) {
+    if (retryCount >= MAX_ERROR_RETRY_COUNT) return;
+
+    const timeout = Math.pow(2, retryCount) * ERROR_RETRY_INTERVAL;
+
+    setTimeout(() => {
+      void action();
+    }, timeout);
```

Comment:
One thing to be aware of with this is that if you schedule a future event like this, there may be some concurrency bugs down the road. Maybe these don't apply here (I've not completely followed all the implementation details here), but suppose `timeout` is 10 seconds at one point and this retry gets scheduled.

Then, say, after 2 seconds, the room gets switched, or the user disconnects, or the user manually retries and that works. Then the user is happy for about 8 seconds, after which this already-scheduled retry kicks in.

Now again, this may not be an issue if we're lucky, but it could potentially become a source of weird issues like "Huh? I disconnected, and after 8 seconds my data pops up again.", or "Huh? I just switched rooms, and now I'm seeing my old rooms data suddenly again?", etc etc. (Sorry if these examples make no sense in this context.)

I went through this kind of pain in our old client before, which is why we've built the state machine for the 1.1 release, finally putting those issues behind us.

What's missing when setting things up like that is a way to manually cancel such an already-scheduled timeout in hindsight. You could do that the simple way, by tracking a `cancelled` boolean in this closure (or maybe in the `subscribersByQuery` mapping as I think you'll only ever want to run a single fetch or retry per query, right?), and checking it when the timeout fires, just to see if you're still interested in retrying it.

Alternatively, you could use an AbortController, which allows you to explicitly clear any such timeout. I personally really like that API, as it's flexible and allows you to cancel an outstanding promise or timeout at any given time or place.

Let me know if I can help in this area in any way.

Reply:
This should not be an issue in our particular situation but you do have me thinking whether it makes sense to cancel the fetch request or timeout if the room is disconnected. Not doing so shouldn't hurt at all, but cancelling the outstanding promises and timeout is more efficient. However, it will increase the code complexity. I want to spend some more time thinking about this and maybe create another PR in future to make these additions. Thanks for offering to help @nvie, and I will definitely be seeking your help when implementing this.